### PR TITLE
[pruner] add cuda tests for pruner

### PR DESCRIPTION
**Pull request by kazhou**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61993

Repeating `test_pruner` unit tests for Linear and Conv2d models with device = 'cuda' to confirm pruner will work on GPU
- set device to cuda
- move model to device
- assert that module.weight.device is cuda

Differential Revision: [D29829293](https://our.internmc.facebook.com/intern/diff/D29829293/)

### CODE REVIEW

File: test/ao/sparsity/test_pruner.py [hunk 1]
```diff
@@ -150,6 +171,25 @@ def test_prepare_linear(self):
             assert type(module.parametrizations.weight[0]) == PruningParametrization
         assert model(x).shape == (128, 16)
 
+    def test_prepare_linear_cuda(self):
```
Comment (supriyar) [inline test/ao/sparsity/test_pruner.py:174]: Do we need separate tests for all the `cuda` cases or can we parametrize the existing one to work on both devices?

**Review by supriyar — CHANGES_REQUESTED** [review]
I think we can try to avoid the code duplication here

File: test/ao/sparsity/test_pruner.py (hunk 1 above)
Comment (z-a-f) [reply test/ao/sparsity/test_pruner.py:174]: I agree -- we can wrap the test in a separate function:

```
def _test_prepare_linear_on_device(self, device):
    print(f'Testing on {device.type}')
    # All of your testing code here

def test_prepare_linear(self):
    self._test_prepare_linear_on_device(torch.device('cpu'))
    if torch.cuda.is_available():
        self._test_prepare_linear_on_device(torch.device('cuda'))
```

**Review by z-a-f — CHANGES_REQUESTED** [review]
some comments

File: test/ao/sparsity/test_pruner.py [hunk 2]
```diff
@@ -155,24 +140,14 @@ def test_constructor_cuda(self):
...
-            # Check parametrization exists and is correct
-            assert parametrize.is_parametrized(module)
-            assert hasattr(module, "parametrizations")
-            # Assume that this is the 1st/only parametrization
-            assert type(module.parametrizations.weight[0]) == PruningParametrization
-        assert model(x).shape == (128, 16)
+    def test_constructor(self):
+        self._test_constructor_on_device(torch.device('cpu'))
+        if torch.cuda.is_available():
+            self._test_constructor_on_device(torch.device('cuda'))
```
Comment (supriyar) [inline test/ao/sparsity/test_pruner.py:146]: Can we use https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_workflow_module.py#L674 directly?

File: test/ao/sparsity/test_pruner.py [hunk 3]
```diff
@@ -118,25 +118,10 @@ def update_mask(self, layer, **kwargs):
...
-        assert len(pruner.module_groups) == 1
-        assert pruner.module_groups[0]['path'] == 'linear'
-        assert 'test' in pruner.module_groups[0]
-        assert pruner.module_groups[0]['test'] == 3
 
-    def test_constructor_cuda(self):
-        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-        # Cannot instantiate the base
+    def _test_constructor_on_device(self, device):
+        print(f'Testing on {device.type}')
```
Comment (supriyar) [inline test/ao/sparsity/test_pruner.py:123]: Please remove print statements

File: test/ao/sparsity/test_pruner.py (hunk 2 above)
Comment (jerryzh168) [reply test/ao/sparsity/test_pruner.py:146]: yeah, also probably just loop through a list of devices and don't use hypothesis.

File: test/ao/sparsity/test_pruner.py [hunk 4]
```diff
@@ -182,136 +213,201 @@ def test_prepare_linear_bias(self):
...
             assert not hasattr(module, 'mask')
         assert model(x).shape == (128, 16)
 
-    def test_convert_linear_bias(self):
+    def test_convert_linear(self):
+        self._test_convert_linear_on_device(torch.device('cpu'))
+        if torch.cuda.is_available():
+            self._test_convert_linear_on_device(torch.device('cuda'))
+
+    def _test_convert_linear_bias_on_device(self, device):
```
Comment (jerryzh168) [inline test/ao/sparsity/test_pruner.py:241]: it looks like there is still a lot of duplications in these code, is it possible to simplify further?

File: test/ao/sparsity/test_pruner.py [hunk 5]
```diff
@@ -118,13 +121,49 @@ def update_mask(self, layer, **kwargs):
...
+            module = g['module']
+            assert module.weight.device == device
+            # Check mask exists
+            assert hasattr(module, 'mask')
+            # Check parametrization exists and is correct
+            assert parametrize.is_parametrized(module)
+            assert hasattr(module, "parametrizations")
+            # Assume that this is the 1st/only parametrization
+            assert type(module.parametrizations.weight[0]) == PruningParametrization
+        return True
```
Comment (jerryzh168) [inline test/ao/sparsity/test_pruner.py:135]: nit: maybe rename this to something like _check_prepared_pruner` and remove the return, similar for other functions

**Review by jerryzh168 — APPROVED** [review]
Thanks for deduplicating the code!

**Pull request closed without merging**